homeassistant/components/humidifier/__init__.py

```diff
@@ -0,0 +1,312 @@
+    @property
+    def humidifier_action(self) -> Optional[str]:
+        """Return the current running humidifier operation if supported.
+
+        Need to be one of CURRENT_HUMIDIFIER_*.
+        """
+        return None
+
+    @property
+    def preset_mode(self) -> Optional[str]:
+        """Return the current preset mode, e.g., home, away, temp.
+
+        Requires SUPPORT_PRESET_MODE.
+        """
+        raise NotImplementedError
+
+    @property
+    def preset_modes(self) -> Optional[List[str]]:
+        """Return a list of available preset modes.
+
+        Requires SUPPORT_PRESET_MODE.
+        """
+        raise NotImplementedError
+
+    def set_humidity(self, humidity: int) -> None:
+        """Set new target humidity."""
+        raise NotImplementedError()
+
+    async def async_set_humidity(self, humidity: int) -> None:
+        """Set new target humidity."""
+        await self.hass.async_add_executor_job(self.set_humidity, humidity)
+
+    def set_operation_mode(self, operation_mode: str) -> None:
+        """Set new target operation mode."""
+        raise NotImplementedError()
+
+    async def async_set_operation_mode(self, operation_mode: str) -> None:
+        """Set new target operation mode."""
+        await self.hass.async_add_executor_job(self.set_operation_mode, operation_mode)
+
+    def set_fan_mode(self, fan_mode: str) -> None:
+        """Set new target fan mode."""
+        raise NotImplementedError()
+
+    async def async_set_fan_mode(self, fan_mode: str) -> None:
+        """Set new target fan mode."""
+        await self.hass.async_add_executor_job(self.set_fan_mode, fan_mode)
+
+    def set_preset_mode(self, preset_mode: str) -> None:
+        """Set new preset mode."""
+        raise NotImplementedError()
+
+    async def async_set_preset_mode(self, preset_mode: str) -> None:
+        """Set new preset mode."""
+        await self.hass.async_add_executor_job(self.set_preset_mode, preset_mode)
+
+    async def async_turn_on(self) -> None:
+        """Turn the entity on."""
+        if hasattr(self, "turn_on"):
+            # pylint: disable=no-member
+            await self.hass.async_add_executor_job(self.turn_on)
+            return
+
+        # Fake turn on
+        for mode in (
+            OPERATION_MODE_HUMIDIFY_DRY,
+            OPERATION_MODE_HUMIDIFY,
+            OPERATION_MODE_DRY,
+        ):
+            if mode not in self.operation_modes:
+                continue
+            await self.async_set_operation_mode(mode)
+            break
+
+    async def async_turn_off(self) -> None:
+        """Turn the entity off."""
+        if hasattr(self, "turn_off"):
+            # pylint: disable=no-member
+            await self.hass.async_add_executor_job(self.turn_off)
+            return
+
+        # Fake turn off
+        if OPERATION_MODE_OFF in self.operation_modes:
+            await self.async_set_operation_mode(OPERATION_MODE_OFF)
+
+    @property
+    def supported_features(self) -> int:
```

Comment:
Should this be an abstract method? Otherwise we need a default, right?

Reply:
Ok, I think SUPPORT_TARGET_HUMIDITY would be a good default.
Should I add defaults or convert to abstract method all functions raising NotImplementedError?

Reply:
I think that target humidity should not be a feature but should be standard and we should not allow humidifiers that cannot humidify…

Reply:
I agree

Reply:
What if they can humidify but can't stop by themselves when reach a certain humidity level?

Reply:
I think that should not be decided upon at this point, define it abstract and leave feature sets up to the integration. 

A dump humidifier with a smart plug, is still a smart plug, thus a switch?

Reply:
Ok, defined it as abstract class and removed the SUPPORT_TARGET_HUMIDITY flag.